File: packages/codex-session-relay/src/codex_session_relay/transport.py
"""Turning one transport receipt into delivery facts.

The ordering matters more than the table. A busy refusal and an approval-policy refusal both
arrive as an RPC error whose method prefix would misclassify them, so the error CODE is read
first. And an unfinished receipt is never proof of non-delivery: the transport writes a
receipt before its calls run, so a missing field can simply mean it has not got there yet.
"""

from dataclasses import dataclass, field

KNOWN_METHODS = ("initialize", "thread/read", "thread/resume", "turn/start")

# Refusals decided from a resume response, before any turn/start. Each keeps its own code so an
# operator sees which one fired; they share one classification because they share one fact:
# nothing was sent. An UNRECOGNISED code is deliberately not in this set - an unknown refusal is
# not a known non-delivery, and it must stay uncertain.
SETTINGS_REFUSALS = (
    "settings_not_preserved",
    # The host reported no value for a requested setting. Decided from the resume response before
    # any turn/start, so nothing was sent: it belongs with the completed pre-send refusals. An
    # ABSENT approval policy therefore lands here rather than in inbox_only, because not seeing a
    # policy is not the same as seeing an interactive one and does not prove the push channel is
    # permanently closed. A REPORTED non-never policy is unchanged and still inbox_only.
    "setting_unobservable",
    "environments_unknown",
    "unverifiable_permission_profile",
)

ACCEPTED = "accepted"
FAILED = "failed"
OUTCOME_UNKNOWN = "outcome_unknown"
UNFINISHED = "in_progress_or_unknown"

DISPATCHED = "dispatched"
DEFERRED_BUSY = "deferred_busy"
WITHHELD_PRE_SEND = "withheld_pre_send"
HELD_UNCERTAIN = "held_uncertain"
INBOX_ONLY = "inbox_only"
QUEUED = "queued"
ACKNOWLEDGED = "acknowledged"
SUPERSEDED = "superseded"


@dataclass(frozen=True)
class TransportFacts:
    transport_receipt_status: str
    delivery_state: str
    send_attempted: str
    retry_safe: bool
    failed_operation: str | None
    turn_id: str | None = None
    approval_policy: str | None = None
    rpc_error_code: str | None = None
    error_text: str | None = None
    notes: tuple = field(default_factory=tuple)


def method_prefix(error_text) -> str | None:
    """RPC errors render as '<method>: <message>'. Anything else is not a method."""
    if not isinstance(error_text, str) or ": " not in error_text:
        return None
    candidate = error_text.split(": ", 1)[0]
    return candidate if candidate in KNOWN_METHODS else None


def _usable_turn_id(value) -> str | None:
    # A dispatched record needs a non-empty turn id; null or blank is not a dispatch.
    return value if isinstance(value, str) and value.strip() else None


def classify_operation_receipt(receipt) -> TransportFacts:
    if not isinstance(receipt, dict):
        return TransportFacts(OUTCOME_UNKNOWN, HELD_UNCERTAIN, "unknown", False, "unclassified")
    status = receipt.get("status")
    error_text = receipt.get("error")
    rpc_error = receipt.get("rpcError") or {}
    code = rpc_error.get("code")
    resumed = receipt.get("resumed")
    policy = (resumed or {}).get("approvalPolicy") if isinstance(resumed, dict) else None
    turn_id = _usable_turn_id(receipt.get("turnId"))

    if status == ACCEPTED:
        if turn_id:
            return TransportFacts(
                ACCEPTED, DISPATCHED, "yes", False, None, turn_id, policy, code, error_text
            )
        # Defensive: the current transport always sets a turn id before reporting acceptance,
        # so this is unreachable through it. Refusing an inconsistent receipt is still better
        # than inventing a dispatch that has no turn to point at.
        return TransportFacts(
            OUTCOME_UNKNOWN, HELD_UNCERTAIN, "unknown", False, "unclassified", None, policy,
            code, error_text, ("accepted receipt carried no usable turn id",),
        )
    if status == UNFINISHED:
        return TransportFacts(
            UNFINISHED, HELD_UNCERTAIN, "unknown", False, None, None, policy, code, error_text
        )
    if status == OUTCOME_UNKNOWN:
        return TransportFacts(
            OUTCOME_UNKNOWN, HELD_UNCERTAIN, "unknown", False, "transport", None, policy,
            code, error_text,
        )
    if status == FAILED:
        if code == "thread_busy" and not resumed:
            return TransportFacts(
                FAILED, DEFERRED_BUSY, "no", True, "thread/read", None, policy, code, error_text
            )
        if code == "unsupported_approval_policy" and resumed:
            return TransportFacts(
                FAILED, INBOX_ONLY, "no", False, "thread/resume", None, policy, code, error_text
            )
        if code in SETTINGS_REFUSALS and resumed:
            # A settings refusal is decided FROM the resume response, so resumed is present and
            # the generic pre-send branch below (which requires resumed to be absent) would miss
            # it. Left unmapped these fall through to held_uncertain, which is the worst answer
            # available: nothing was started, yet the attempt would be parked as possibly
            # delivered and reconciliation could never establish non-delivery from the receipt.
            #
            # retrySafe is true strictly as duplicate-delivery safety: this is a completed,
            # attributable local refusal before turn/start. It asserts nothing about whether the
            # resume itself changed host state.
            return TransportFacts(
                FAILED, WITHHELD_PRE_SEND, "no", True, "thread/resume", None, policy, code,
                error_text,
            )
        prefix = method_prefix(error_text)
        if prefix == "initialize":
            return TransportFacts(
                FAILED, HELD_UNCERTAIN, "unknown", False, "initialize", None, policy, code,
                error_text,
            )
        if prefix == "turn/start":
            return TransportFacts(
                FAILED, HELD_UNCERTAIN, "yes", False, "turn/start", None, policy, code, error_text
            )
        if prefix in ("thread/read", "thread/resume") and not resumed:
            return TransportFacts(
                FAILED, WITHHELD_PRE_SEND, "no", True, prefix, None, policy, code, error_text
            )
    return TransportFacts(
        OUTCOME_UNKNOWN, HELD_UNCERTAIN, "unknown", False, "unclassified", None, policy, code,
        error_text, ("receipt shape not recognised",),
    )


def attempt_record(facts: TransportFacts, *, request_id, event_id, attempt_no, recipient,
                   status_before, observed_at, reconciliation=None) -> dict:
    """The frozen DeliveryAttempt shape. Relay-internal notes never appear here."""
    record = {
        "requestId": request_id,
        "eventId": event_id,
        "attemptNo": attempt_no,
        "recipientTaskId": recipient,
        "deliveryState": facts.delivery_state,
        "sendAttempted": facts.send_attempted,
        "retrySafe": facts.retry_safe,
        "observedAt": observed_at,
        "recipientStatusBefore": status_before,
        "recipientApprovalPolicy": facts.approval_policy,
        "transportReceiptStatus": facts.transport_receipt_status,
        "failedOperation": facts.failed_operation,
        "turnId": facts.turn_id,
    }
    if reconciliation is not None:
        record["reconciliation"] = reconciliation
    assert_attempt_invariants(record)
    return record


def assert_attempt_invariants(record: dict) -> None:
    """Re-implement the frozen schema's conditional rules, in Python, before persisting."""
    state = record["deliveryState"]
    if state == HELD_UNCERTAIN:
        assert record["retrySafe"] is False, "an uncertain attempt is never retry-safe"
        assert record["sendAttempted"] in ("unknown", "yes")
    if record["retrySafe"]:
        assert record["sendAttempted"] == "no"
        assert record["transportReceiptStatus"] == FAILED
        assert state in (WITHHELD_PRE_SEND, DEFERRED_BUSY)
        assert record["failedOperation"] in ("thread/read", "thread/resume")
    if record["transportReceiptStatus"] == UNFINISHED:
        assert record["sendAttempted"] == "unknown"
        assert record["retrySafe"] is False
    if state == DISPATCHED:
        assert isinstance(record["turnId"], str) and record["turnId"]
        assert record["sendAttempted"] == "yes"
        assert record["retrySafe"] is False
    if record["failedOperation"] in ("turn/start", "transport", "initialize", "unclassified"):
        assert record["retrySafe"] is False
        assert record["sendAttempted"] in ("unknown", "yes")

File: internal/relay/delivery/transport.go
package delivery

import (
	"fmt"
	"slices"
	"strings"
)

// Transport vocabulary (transport.py). These spellings are stored in the database and read by
// every status surface.
const (
	Accepted       = "accepted"
	FailedStatus   = "failed"
	OutcomeUnknown = "outcome_unknown"
	Unfinished     = "in_progress_or_unknown"

	Dispatched      = "dispatched"
	DeferredBusy    = "deferred_busy"
	WithheldPreSend = "withheld_pre_send"
	HeldUncertain   = "held_uncertain"
	InboxOnly       = "inbox_only"
	Queued          = "queued"
	Acknowledged    = "acknowledged"
	Superseded      = "superseded"
	Sending         = "sending"
)

var knownMethods = []string{"initialize", "thread/read", "thread/resume", "turn/start"}

// SettingsRefusals are the resume refusals decided before any turn/start (transport.py).
var SettingsRefusals = []string{"settings_not_preserved", "setting_unobservable", "environments_unknown", "unverifiable_permission_profile", "settings_differ_after_load"}

// Facts is transport.TransportFacts: one receipt turned into delivery facts.
type Facts struct {
	ReceiptStatus   string
	DeliveryState   string
	SendAttempted   string
	RetrySafe       bool
	FailedOperation any // string or nil
	TurnID          any // string or nil
	ApprovalPolicy  any
	RPCErrorCode    any
	ErrorText       any
}

func methodPrefix(errorText any) (string, bool) {
	text, ok := errorText.(string)
	if !ok || !strings.Contains(text, ": ") {
		return "", false
	}
	candidate, _, _ := strings.Cut(text, ": ")
	return candidate, slices.Contains(knownMethods, candidate)
}

func usableTurnID(v any) any {
	if s, ok := v.(string); ok && strings.TrimSpace(s) != "" {
		return s
	}
	return nil
}

func truthy(v any) bool {
	switch t := v.(type) {
	case nil:
		return false
	case bool:
		return t
	case string:
		return t != ""
	case Obj:
		return len(t) > 0
	case []any:
		return len(t) > 0
	case int64:
		return t != 0
	case float64:
		return t != 0
	}
	return true
}

// Classify is transport.classify_operation_receipt. The error CODE is read before the method
// prefix, and an unfinished receipt is never proof of non-delivery.
func Classify(receipt Obj) Facts {
	if receipt == nil {
		return Facts{OutcomeUnknown, HeldUncertain, "unknown", false, "unclassified", nil, nil, nil, nil}
	}
	status, _ := get(receipt, "status")
	errorText, _ := get(receipt, "error")
	rpc, _ := get(receipt, "rpcError")
	var code any
	if o, ok := rpc.(Obj); ok {
		code, _ = get(o, "code")
	}
	resumed, _ := get(receipt, "resumed")
	var policy any
	if o, ok := resumed.(Obj); ok {
		policy, _ = get(o, "approvalPolicy")
	}
	rawTurn, _ := get(receipt, "turnId")
	turn := usableTurnID(rawTurn)
	f := func(rs, ds, sa string, safe bool, op any, turn any) Facts {
		return Facts{rs, ds, sa, safe, op, turn, policy, code, errorText}
	}
	switch status {
	case Accepted:
		if turn != nil {
			return f(Accepted, Dispatched, "yes", false, nil, turn)
		}
		return f(OutcomeUnknown, HeldUncertain, "unknown", false, "unclassified", nil)
	case Unfinished:
		return f(Unfinished, HeldUncertain, "unknown", false, nil, nil)
	case OutcomeUnknown:
		return f(OutcomeUnknown, HeldUncertain, "unknown", false, "transport", nil)
	case FailedStatus:
		if code == "thread_busy" && !truthy(resumed) {
			return f(FailedStatus, DeferredBusy, "no", true, "thread/read", nil)
		}
		if code == "unsupported_approval_policy" && truthy(resumed) {
			return f(FailedStatus, InboxOnly, "no", false, "thread/resume", nil)
		}
		if c, ok := code.(string); ok && slices.Contains(SettingsRefusals, c) && truthy(resumed) {
			return f(FailedStatus, WithheldPreSend, "no", true, "thread/resume", nil)
		}
		prefix, known := methodPrefix(errorText)
		switch {
		case known && prefix == "initialize":
			return f(FailedStatus, HeldUncertain, "unknown", false, "initialize", nil)
		case known && prefix == "turn/start":
			return f(FailedStatus, HeldUncertain, "yes", false, "turn/start", nil)
		case known && (prefix == "thread/read" || prefix == "thread/resume") && !truthy(resumed):
			return f(FailedStatus, WithheldPreSend, "no", true, prefix, nil)
		}
	}
	return f(OutcomeUnknown, HeldUncertain, "unknown", false, "unclassified", nil)
}

// AttemptRecord is transport.attempt_record: the frozen DeliveryAttempt shape.
func AttemptRecord(facts Facts, requestID, eventID string, attemptNo int64, recipient, statusBefore, observedAt string, reconciliation Obj) (Obj, error) {
	record := Obj{
		{Key: "requestId", Value: requestID},
		{Key: "eventId", Value: eventID},
		{Key: "attemptNo", Value: attemptNo},
		{Key: "recipientTaskId", Value: recipient},
		{Key: "deliveryState", Value: facts.DeliveryState},
		{Key: "sendAttempted", Value: facts.SendAttempted},
		{Key: "retrySafe", Value: facts.RetrySafe},
		{Key: "observedAt", Value: observedAt},
		{Key: "recipientStatusBefore", Value: statusBefore},
		{Key: "recipientApprovalPolicy", Value: facts.ApprovalPolicy},
		{Key: "transportReceiptStatus", Value: facts.ReceiptStatus},
		{Key: "failedOperation", Value: facts.FailedOperation},
		{Key: "turnId", Value: facts.TurnID},
	}
	if reconciliation != nil {
		record = append(record, F{Key: "reconciliation", Value: reconciliation})
	}
	return record, AssertAttemptInvariants(record)
}

// AssertAttemptInvariants re-implements the frozen schema's conditional rules before persisting.
func AssertAttemptInvariants(record Obj) error {
	state := str(record, "deliveryState")
	safe, _ := get(record, "retrySafe")
	sent := str(record, "sendAttempted")
	status := str(record, "transportReceiptStatus")
	op, _ := get(record, "failedOperation")
	turn, _ := get(record, "turnId")
	fail := func(rule string) error { return fmt.Errorf("delivery: attempt invariant: %s", rule) }
	if state == HeldUncertain && (safe != false || (sent != "unknown" && sent != "yes")) {
		return fail("an uncertain attempt is never retry-safe")
	}
	if safe == true {
		if sent != "no" || status != FailedStatus || (state != WithheldPreSend && state != DeferredBusy) || (op != "thread/read" && op != "thread/resume") {
			return fail("a retry-safe attempt is a completed pre-send failure")
		}
	}
	if status == Unfinished && (sent != "unknown" || safe != false) {
		return fail("an unfinished receipt is unknown and unsafe")
	}
	if state == Dispatched {
		if t, ok := turn.(string); !ok || t == "" || sent != "yes" || safe != false {
			return fail("a dispatch carries its turn")
		}
	}
	switch op {
	case "turn/start", "transport", "initialize", "unclassified":
		if safe != false || (sent != "unknown" && sent != "yes") {
			return fail("a post-send failure is never retry-safe")
		}
	}
	return nil
}
